app/js/app/directives/equation_editor/EquationInput.js

```diff
@@ -31,29 +88,90 @@ define(["/partials/equation_editor/equation_input.html"], function(templateUrl)
                     }
                     timer = $timeout(function() {
                         let pycode = element.find(".eqn-text-input")[0].value;
-                        let openBracketsCount = pycode.split('(').length - 1;
-                        let closeBracketsCount = pycode.split(')').length - 1;
 
                         scope.state = {result: {python: pycode}, textEntry: true};
-                        let regexStr = "[^ (-)*-/0-9<->A-Z^-_a-z±²-³¼-¾×÷]+";
-                        let badCharacters = RegExp(regexStr);
-                        let goodCharacters = RegExp(regexStr.replace("^", ""), 'g');
                         scope.textEntryError = [];
-                        if (/\\[a-zA-Z()]|[{}]/.test(pycode)) {
-                            scope.textEntryError.push('LaTeX syntax is not supported.');
-                        }
-                        if (badCharacters.test(pycode)) {
-                            scope.textEntryError.push('Some of the characters you are using are not allowed: ' + _.uniq(pycode.replace(goodCharacters, '')).join(' '));
-                        }
-                        if (openBracketsCount != closeBracketsCount) {
-                            scope.textEntryError.push('You are missing some ' + (closeBracketsCount > openBracketsCount ? 'opening' : 'closing') + ' brackets.');
-                        }
-                        if (/\.[0-9]/.test(pycode)) {
-                            scope.textEntryError.push('Please convert decimal numbers to fractions.');
+
+                        let parsedExpression = parseExpression(pycode);
+                        console.log(parsedExpression);
+
+                        if (!parsedExpression.hasOwnProperty('error')) {
+                            if (parsedExpression.length === 0) {
+                                if (pycode === '') {
+                                   element.find(".eqn-preview").html("Click here to enter a formula!");
+                                    scope.symbols = [];
+                                    scope.state.result.tex = "";
+                                    scope.state.result.python = "";
+                                    scope.state.result.mathml = "";
+                                    sketch.symbols = [];
+                                } else {
+                                    // TODO: Do something here
+                                    console.log("User entered maths, we could not parse it.")
```

Comment:
Surely one of the biggest causes of parse errors will be the syntax errors we used to check for immediately (but are now checked down after successful parsing on L122-L137), and so that checking should be done before parsing?